File: 2020/day4/python/day4.py
#!/usr/bin/env python3
from os.path import dirname
from os.path import realpath
from os.path import join
import re
import time
import pytest


def transform_input(input_):
    # custom transform for the day

    passports, n_passports = [dict()], 0

    for inp in input_.splitlines():
        if not inp:
            passports.append(dict())
            n_passports += 1
        else:
            for parts in inp.split():
                key, value = parts.split(":")
                passports[n_passports][key] = value

    return passports


def read_input(file_name="../data/input.txt"):

    dir_path = dirname(realpath(__file__))
    with open(join(dir_path, file_name), "r") as f:
        input_ = transform_input(f.read())

    return input_


def solve_part1(input_):
    # Required fields
    req = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    # Check if all keys for each passport contain the required keys
    return sum([1 for inp in input_ if set(inp).issuperset(set(req))])


def solve_part2(input_):

    # Validation regex (nice idea from user https://github.com/hoxovic)
    rgx = {
        "byr": r"^(19[2-9]\d|200[0-2])$",
        "iyr": r"^20(1\d|20)",
        "eyr": r"^20(2\d|30)",
        "hgt": r"^(1([5-8]\d|9[0-3])cm|(59|6\d|7[0-6])in)$",
        "hcl": r"^#[\da-f]{6}",
        "ecl": r"^(amb|blu|brn|gry|grn|hzl|oth)$",
        "pid": r"^(\d{9})$",
        "cid": r"^",
    }

    def matches(inp):
        return sum(
            [1 for key, val in inp.items() if re.search(rgx[key], val)]
        ) == len(inp)

    return sum([1 for inp in input_ if solve_part1([inp]) and matches(inp)])


def main():

    input_ = read_input("../data/input.txt")
    t0 = time.time()
    part1 = solve_part1(input_)
    time_part1 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part1} (time taken {time_part1}[ms])")

    t0 = time.time()
    part2 = solve_part2(input_)
    time_part2 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part2} (time taken {time_part2}[ms])")


if __name__ == "__main__":
    main()


@pytest.mark.parametrize("input1,output1", [("../data/test_input0.txt", 2)])
def test1(input1, output1):
    assert solve_part1(read_input(input1)) == output1


@pytest.mark.parametrize(
    "input2,output2",
    [
        ("../data/test_input1.txt", 0),
        ("../data/test_input2.txt", 4),
    ],
)
def test2(input2, output2):
    assert solve_part2(read_input(input2)) == output2

File: 2020/day4/cpp/src/day4.cpp
#include "2020/day4/cpp/include/day4.hpp"

#include <regex>
#include <set>

#include "utils/cpp/include/utils.hpp"

using day4::input_type;

input_type day4::transform_input(std::vector<std::string> &input) {
  input_type passports({std::map<std::string, std::string>()});
  int n_passports(0);
  for (auto &inp : input) {
    if (inp.size() == 0) {
      passports.push_back(std::map<std::string, std::string>());
      n_passports++;
    } else {
      std::vector<std::string> parts;
      utils::split_string(inp, ' ', parts);
      for (auto &part : parts) {
        std::vector<std::string> key_value;
        utils::split_string(part, ':', key_value);
        passports[n_passports][key_value[0]] = key_value[1];
      }
    }
  }
  return passports;
}

input_type day4::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

// Implement part one solution
std::string day4::solve_part1(const input_type &input_) {
  // Required fields
  std::string req[] = {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};
  std::set<std::string> req_set(std::begin(req), std::end(req));

  auto n_matches(0);
  for (auto &inp : input_) {
    std::set<std::string> inp_set;
    for (auto &i : inp) {
      inp_set.insert(i.first);
    }

    if (std::includes(inp_set.begin(), inp_set.end(), req_set.begin(),
                      req_set.end())) {
      n_matches++;
    }
  }
  return std::to_string(n_matches);
}

// Implement part two solution
std::string day4::solve_part2(const input_type &input_) {
  std::map<std::string, std::string> rgx = {
      {"byr", "^(19[2-9]\\d|200[0-2])$"},
      {"iyr", "^20(1\\d|20)"},
      {"eyr", "^20(2\\d|30)"},
      {"hgt", "^(1([5-8]\\d|9[0-3])cm|(59|6\\d|7[0-6])in)$"},
      {"hcl", "^#[\\da-f]{6}"},
      {"ecl", "^(amb|blu|brn|gry|grn|hzl|oth)$"},
      {"pid", "^(\\d{9})$"},
      {"cid", "^(.*)"}};

  int n_matches(0);
  for (auto &inp : input_) {
    size_t rgx_matches(0);
    for (auto &mp : inp) {
      rgx_matches +=
          (std::regex_match(mp.second, std::regex(rgx[mp.first]))) ? 1 : 0;
    }
    n_matches +=
        (rgx_matches == inp.size() && solve_part1({inp}) == "1") ? 1 : 0;
  }
  return std::to_string(n_matches);
}